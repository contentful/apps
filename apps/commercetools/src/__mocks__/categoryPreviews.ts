import { Category } from './../interfaces';

const categoryPreviews: Category[] = [
  {
    id: '28a7442d',
    name: "Mens' clothes",
    slug: 'men-regular',
    externalLink: 'https://mc.commercetools.com/example-project/categories/28a7442d/general'
  },
  {
    id: '38a7442d',
    name: "Womens' clothes",
    slug: 'women-regular',
    externalLink: 'https://mc.commercetools.com/example-project/categories/38a7442d/general'
  },
  {
    id: '48a7442d',
    name: "Kids' clothes",
    slug: 'kids-regular',
    externalLink: 'https://mc.commercetools.com/example-project/categories/48a7442d/general'
  }
];

export default categoryPreviews;
