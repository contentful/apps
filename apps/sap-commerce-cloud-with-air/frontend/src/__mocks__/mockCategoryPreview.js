export const mockCategoryPreview = {
  sku: 'abc1234',
  image: '',
  id: '123',
  name: 'Mock Product',
  slug: '',
  isMissing: false,
};
