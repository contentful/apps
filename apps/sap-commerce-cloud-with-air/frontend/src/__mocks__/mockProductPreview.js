export const mockProductPreview = {
  sku: 'abc1234',
  image: '',
  id: '123',
  name: 'Mock Category',
  isMissing: false,
};
