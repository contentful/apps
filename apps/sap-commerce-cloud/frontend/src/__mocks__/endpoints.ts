export const mockApiEndpoint = 'localhost:9002';
export const mockBaseSite = 'electronics-spa';
