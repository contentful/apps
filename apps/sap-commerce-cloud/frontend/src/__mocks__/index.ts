export { makeSdkMock } from './mockSdk';
export { mockProductPreview } from './mockProductPreview';
