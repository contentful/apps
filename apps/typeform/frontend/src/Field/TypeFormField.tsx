import React, { useEffect, useReducer } from 'react';
import { FieldExtensionSDK, AppExtensionSDK } from '@contentful/app-sdk';
import { Select, Option, TextLink, Note, Tooltip } from '@contentful/forma-36-react-components';
import { TypeFormResponse, FormOption, InstallationParameters } from '../typings';
import { TypeformOAuth } from '../Auth/TypeformOAuth';
import { styles } from './styles';
// @ts-ignore 2307
import logo from './typeform-icon.svg';
import { isUserAuthenticated, getToken, resetLocalStorage } from '../utils';


interface Props {
  sdk: FieldExtensionSDK & AppExtensionSDK;
}

enum ACTION_TYPES {
  INIT = 'INIT',
  UPDATE_VALUE = 'UPDATE_VALUE',
  UPDATE_TOKEN = 'UPDATE_TOKEN',
  RESET = 'RESET',
  ERROR = 'ERROR'
}

const initialState = {
  error: false,
  value: '',
  selectedForm: {
    name: '',
    href: '',
    isPublic: true,
    id: ''
  } as FormOption,
  hasStaleData: false,
  token: getToken(),
  forms: [] as FormOption[],
  loading: true
};
const AUTH_ERROR_CODES = [401, 403];

const isStaleData = (value: string, forms: FormOption[]): boolean => {
  if (value) {
    if (forms.length === 0) {
      return true;
    } else {
      if (forms.find(form => form.href === value)) {
        return false;
      } else {
        return true;
      }
    }
  } else {
    return false;
  }
};

const getSelectedForm = (value: string, forms: FormOption[]) => {
  return forms.find(form => form.href === value) || initialState.selectedForm;
};

export function TypeFormField({ sdk }: Props) {
  const { selectedWorkspaceId } = sdk.parameters.installation as InstallationParameters;
  const [state, dispatch] = useReducer(reducer, initialState);
  const { loading, forms, value, hasStaleData, selectedForm, error, token } = state;

  function reducer(
    state = initialState,
    action: { type: string; payload?: any }
  ): typeof initialState {
    switch (action.type) {
      case ACTION_TYPES.INIT: {
        const { forms } = action.payload;
        const currentFieldValue = sdk.field.getValue();
        const hasStaleData = isStaleData(currentFieldValue, forms);
        return {
          ...state,
          value: currentFieldValue,
          selectedForm: getSelectedForm(currentFieldValue, forms),
          loading: false,
          forms,
          error: false,
          hasStaleData
        };
      }
      case ACTION_TYPES.UPDATE_VALUE: {
        const { value, forms } = action.payload;
        let selectedForm = initialState.selectedForm;
        if (value) {
          sdk.field.setValue(value);
          selectedForm = (forms as FormOption[]).find(form => form.href === value)!;
        } else {
          selectedForm = initialState.selectedForm;
          sdk.field.removeValue();
        }
        return { ...state, value, selectedForm, hasStaleData: false };
      }
      case ACTION_TYPES.RESET: {
        sdk.field.removeValue();
        return {
          ...state,
          value: '',
          hasStaleData: false,
          selectedForm: initialState.selectedForm
        };
      }
      case ACTION_TYPES.UPDATE_TOKEN: {
        const { token } = action.payload;
        return {
          ...state,
          loading: true,
          error: false,
          token
        };
      }
      case ACTION_TYPES.ERROR: {
        return { ...state, loading: false, error: true };
      }
      default:
        return state;
    }
  }

  useEffect(() => {
    const fetchForms = async () => {
      try {
        const response = await fetch(`/forms/${selectedWorkspaceId}`, {
          headers: {
            Authorization: `Bearer ${token}`
          }
        });

        if (AUTH_ERROR_CODES.includes(response.status)) {
          // clear everything in case token is expired
          resetLocalStorage();
          dispatch({ type: ACTION_TYPES.RESET });
        } else {
          const result = (await response.json()) as TypeFormResponse;
          const normalizedForms = normalizeFormResponse(result);
          dispatch({
            type: ACTION_TYPES.INIT,
            payload: {
              forms: normalizedForms
            }
          });
        }
      } catch (error) {
        // only show error dialog is the user is logged in
        if (isUserAuthenticated()) {
          dispatch({ type: ACTION_TYPES.ERROR });
        }
      }
    };
    fetchForms();
    // Start auto resizer to adjust field height
    sdk.window.startAutoResizer();
  }, [token]);

  const onChange = (event: any) => {
    const value = event.currentTarget.value;
    dispatch({ type: ACTION_TYPES.UPDATE_VALUE, payload: { value, forms } });
  };

  const openDialog = async () => {
    await sdk.dialogs.openCurrentApp({
      width: 1000,
      parameters: {
        value
      },
      title: 'Form Preview',
      shouldCloseOnEscapePress: true,
      shouldCloseOnOverlayClick: true
    });
  };

  const normalizeFormResponse = (typeFormResponse: TypeFormResponse): FormOption[] => {
    return typeFormResponse.forms.items.map(form => ({
      name: form.title,
      href: form._links.display,
      id: form.id,
      isPublic: form.settings.is_public
    }));
  };

  if (!isUserAuthenticated()) {
    return (
      <TypeformOAuth
        data-test-id="typeform-auth"
        isFullWidth={false}
        setToken={(token: string) =>
          dispatch({ type: ACTION_TYPES.UPDATE_TOKEN, payload: { token } })
        }
      />
    );
  }

  if (loading) {
    return null;
  }

  if (error) {
    return (
      <Note noteType="negative">
        We could not fetch your typeforms. Please make sure you have selected a valid Typeform
        workspace and are logged in.
      </Note>
    );
  }

  const PreviewButton = (
    <div className={styles.previewButton(!selectedForm.isPublic)}>
      <TextLink onClick={openDialog} disabled={!selectedForm.isPublic}>
        <svg width="16" height="16" viewBox="0 0 24 24">
          <path d="M0 0h24v24H0z" fill="none" />
          <path d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z" />
        </svg>
        Preview
      </TextLink>
    </div>
  );

  return (
    <React.Fragment>
      <div className={styles.field}>
        <img src={logo} className={styles.logo} />
        <Select onChange={onChange} value={value} data-test-id="typeform-select">
          <Option key="" value="">
            {forms.length === 0 ? 'No forms available' : 'Choose a typeform'}
          </Option>
          {forms.map(form => (
            <Option key={form.id} value={form.href}>
              {form.name}
            </Option>
          ))}
        </Select>
      </div>
      {value && !hasStaleData && (
        <div className={styles.actionButtons}>
          <TextLink
            href={`https://admin.typeform.com/form/${selectedForm.id}/create`}
            target="_blank"
            icon="Edit"
            rel="noopener noreferrer"
            className={styles.editButton}
            disabled={!value}>
            Edit
          </TextLink>
          {selectedForm.isPublic ? (
            PreviewButton
          ) : (
            <Tooltip
              containerElement="span"
              content="You can not preview this typeform because it is private"
              place="right">
              {PreviewButton}
            </Tooltip>
          )}
          <TextLink
            href={`https://admin.typeform.com/form/${selectedForm.id}/results`}
            target="_blank"
            icon="Entry"
            rel="noopener noreferrer"
            className={styles.editButton}>
            Results
          </TextLink>
        </div>
      )}
      {hasStaleData && (
        <Note noteType="negative">
          The typeform you have selected in Contentful no longer exists in your workspace.{' '}
          <TextLink onClick={() => dispatch({ type: ACTION_TYPES.RESET })}>Clear field</TextLink>.
        </Note>
      )}
    </React.Fragment>
  );
}
