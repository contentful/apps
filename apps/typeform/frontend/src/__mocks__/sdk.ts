export const sdk: any = {
  ids: {
    space: 'test-space',
    environment: 'master',
    user: '123',
  },
  app: {
    setReady: jest.fn(),
    isInstalled: jest.fn().mockReturnValue(Promise.resolve(false)),
    getParameters: jest.fn().mockReturnValue(Promise.resolve(null)),
    onConfigure: jest.fn(),
  },
  space: {
    getContentTypes: jest.fn().mockReturnValue(Promise.resolve([])),
    getEditorInterfaces: jest.fn().mockResolvedValue({ items: [] }),
  },
  notifier: {
    error: jest.fn(),
  },
  window: {
    startAutoResizer: jest.fn(),
  },
  field: {
    getValue: jest.fn().mockReturnValue('field-value'),
    setValue: jest.fn(),
    removeValue: jest.fn(),
  },
  parameters: {
    installation: {
      workspaceId: 'mock-workspace-id',
      accessToken: 'mock-access-token',
    },
  },
};
