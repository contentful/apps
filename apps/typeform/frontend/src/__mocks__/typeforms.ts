export const typeforms = {
  forms: {
    total_items: 2,
    page_count: 1,
    items: [
      {
        id: 'OhBix6',
        title: 'Customer Feedback Survey (copy)',
        last_updated_at: '2020-03-03T14:07:00Z',
        settings: { is_public: false, is_trial: false },
        self: { href: 'https://api.typeform.com/forms/Oksaodk' },
        theme: { href: 'https://api.typeform.com/themes/fkoaskfd' },
        _links: { display: 'https://user.typeform.com/to/dsako' },
      },
      {
        id: 'sUM1zW',
        title: 'Personality Quiz - Quick Start - FINAL (copy)',
        last_updated_at: '2020-03-03T14:05:55Z',
        settings: { is_public: true, is_trial: true },
        self: { href: 'https://api.typeform.com/forms/dsjao' },
        theme: { href: 'https://api.typeform.com/themes/dsajo' },
        _links: { display: 'https://user.typeform.com/to/sjao' },
      },
    ],
  },
};
