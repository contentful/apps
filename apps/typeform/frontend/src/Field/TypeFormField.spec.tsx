import React from 'react';
import { configure, render, cleanup } from '@testing-library/react';
import { TypeFormField } from './TypeFormField';
import { typeforms } from '../__mocks__/typeforms';
import { sdk as mockSdk } from '../__mocks__/sdk';

configure({
  testIdAttribute: 'data-test-id',
});

describe('TypeFormField', () => {
  beforeEach(() => {
    window.fetch = jest.fn(() => ({
      json: () => typeforms,
    })) as any;
  });
  it('should render successfully when the user is signed in', async () => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => 'some token'),
      },
      writable: true,
    });
    const component = render(<TypeFormField sdk={mockSdk} />);
    await component.findByTestId('typeform-select');
    expect(component.container).toMatchSnapshot();
  });

  it('should render the auth button when the user is not authenticated', async () => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => null),
      },
      writable: true,
    });
    const component = render(<TypeFormField sdk={mockSdk} />);
    await component.findByTestId('typeform-auth');
    expect(component.container).toMatchSnapshot();
  });

  it('should render the error screen if something goes wrong', async () => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => 'token'),
      },
      writable: true,
    });

    window.fetch = jest.fn(() => ({
      json: () => new Error(),
    })) as any;

    const component = render(<TypeFormField sdk={mockSdk} />);
    await component.findByText(/^We could not fetch your typeforms/);
    expect(component.container).toMatchSnapshot();
  });
});
