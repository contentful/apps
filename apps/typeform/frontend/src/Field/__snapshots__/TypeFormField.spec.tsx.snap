// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeFormField should render successfully when the user is signed in 1`] = `
<div>
  <div
    class="css-1cimf61"
  >
    <img
      alt="TypeForm Logo"
      class="css-1ngw63v"
      src="typeform-icon.svg"
    />
    <div
      class="Select__Select__wrapper___2mbYV Select__Select--full___AENS4"
    >
      <select
        class="Select__Select___2Gi9N a11y__focus-border--default___60AXp"
        data-test-id="typeform-select"
      >
        <option
          data-test-id="cf-ui-select-option"
          value=""
        >
          Choose a typeform
        </option>
        <option
          data-test-id="cf-ui-select-option"
          value="https://user.typeform.com/to/dsako"
        >
          Customer Feedback Survey (copy)
        </option>
        <option
          data-test-id="cf-ui-select-option"
          value="https://user.typeform.com/to/sjao"
        >
          Personality Quiz - Quick Start - FINAL (copy)
        </option>
      </select>
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD Select__Select__icon___OBmvS"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M5.29289 8.29289C5.68342 7.90237 6.31658 7.90237 6.70711 8.29289L12 13.5858L17.2929 8.29289C17.6834 7.90237 18.3166 7.90237 18.7071 8.29289C19.0976 8.68342 19.0976 9.31658 18.7071 9.70711L12.7071 15.7071C12.3166 16.0976 11.6834 16.0976 11.2929 15.7071L5.29289 9.70711C4.90237 9.31658 4.90237 8.68342 5.29289 8.29289Z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`TypeFormField should render the auth button when the user is not authenticated 1`] = `
<div>
  <button
    class="Button__Button___1ZfFj Button__Button--primary___JImeO"
    data-test-id="typeform-auth"
    type="button"
  >
    <span
      class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
      tabindex="-1"
    >
      <span
        class="Button__Button__label___3tcOj"
      >
        Sign in to Typeform
      </span>
    </span>
  </button>
</div>
`;

exports[`TypeFormField should render the error screen if something goes wrong 1`] = `
<div>
  <div
    class="Note__Note___2eSKN Note__Note--negative___3fkwL"
    data-test-id="cf-ui-note"
  >
    <div
      class="Note__Note__icon___20RqC"
    >
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--negative___1dled"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
        />
      </svg>
    </div>
    <div
      class="Note__Note__info___2m_-K"
    >
      <div
        class="Note__Note__content___H1pf-"
      >
        We could not fetch your typeforms. Please make sure you have selected a valid Typeform workspace and are logged in.
      </div>
    </div>
  </div>
</div>
`;
